File: 150-evaluate-reverse-polish-notation/leochang/150.cpp
class Solution {
public:
	int evalRPN(vector<string>& tokens) {
		stack<int> tmp;
		for (auto token: tokens) {
			int n1, n2;
			bool isPop = false;
			if (tmp.size() >= 2) {
				n2 = tmp.top();
				tmp.pop();
				n1 = tmp.top();
				tmp.pop();
				isPop = true;
			}
			if (token == "+") {
				tmp.push(n1 + n2);
			} else if (token == "-") {
				tmp.push(n1 - n2);
			} else if (token == "*") {
				tmp.push(n1 * n2);
			} else if (token == "/") {
				tmp.push(n1 / n2);
			} else {
				if (isPop) {
					tmp.push(n1);
					tmp.push(n2);
				}
				tmp.push(stoi(token));
			}
		}

		return tmp.top();
	}
};

File: 150-evaluate-reverse-polish-notation/ken/150.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """

        def isOP(token):
            if token == "+" or \
               token == "-" or \
               token == "*" or \
               token == "/":
               return True
            return False

        def OP(op, a, b):
            res = 0
            if token == "+":
                res = int(b) + int(a)
            elif token == "-":
                res = int(b) - int(a)
            elif token == "*":
                res = int(b) * int(a)
            elif token == "/":
                res = int(float(int(b)) / int(a))
            return res

        stack = []

        while len(tokens) > 0:
            token = tokens.pop(0)

            if isOP(token):
                a = stack.pop()
                b = stack.pop()
                res = OP(token, a, b)
                stack.append(str(res))
            else:
                stack.append(token)

        return int(stack[0])
